File: cpp/euler06.cpp
// Project Euler -- Problem #6
// Solution by nathan dotz aka sleepynate
// nathan (period) dotz (at sign) gmail (period) com
//
// Find the difference between the sum of the squares of the first one
// hundred natural numbers and the square of the sum.

#include <iostream>
using namespace std;

int working_set[100] = {};

bool Fill_working_set() {
  int i = 0;
  while (i<100)
    working_set[i] = ++i;
  return true;
}

int main() {
  if (Fill_working_set())
    cout << "Sample array filled" << endl;
    // find sum of squares
    int i = 0;
    unsigned long int SUM_OF_SQUARES = 0;
    while (i<100) {
      SUM_OF_SQUARES += (working_set[i] * working_set[i]);
      i++;
    }
    cout << SUM_OF_SQUARES << " is the sum of the squares 1..100" << endl;
    // find the square of the sum
    i = 0;
    unsigned long int SQUARE_OF_SUM = 0;
    while (i<100) {
      SQUARE_OF_SUM += working_set[i];
      i++;
    }
    SQUARE_OF_SUM *= SQUARE_OF_SUM;
    cout << SQUARE_OF_SUM << " is the square of the sum" << endl;
    cout << SQUARE_OF_SUM - SUM_OF_SQUARES << " is the difference" << endl;
  return 0;
}

File: python/Euler06.py
# Project Euler - Problem 6

# The sum of the squares of the first ten natural numbers is,
# 1^2 + 2^2 + ... + 10^2 = 385
#
# The square of the sum of the first ten natural numbers is,
# (1 + 2 + ... + 10)^2 = 552 = 3025
# Hence the difference between the sum of the squares of the first ten natural
# numbers and the square of the sum is 3025 − 385 = 2640.
#
# Find the difference between the sum of the squares of the first one hundred
# natural numbers and the square of the sum.

import unittest

class SumOfSquaresTest(unittest.TestCase):
    """Tests for Sum of squares problem"""

    def test_can_find_sum_of_squares(self):
        self.assertEqual(sum_of_squares(1, 10), 385)

    def test_can_find_square_of_sum(self):
        self.assertEqual(square_of_sum(1, 10), 3025)

def sum_of_squares(start, end):
    return sum(map(lambda i: i ** 2, range(start, end + 1)))

def square_of_sum(start, end):
    return sum(range(start, end + 1)) ** 2

if __name__ == '__main__':
    n = abs(sum_of_squares(1, 100) - square_of_sum(1, 100))
    print("The difference between the sum of squares and the square of the sum from 1-100 is ", n)
